Add support for vcf lift-over.

niyarin opened the pull request:
This PR adds varity.vcf-lift/vcf-liftover .
Currently, one coordinate lift-over is supported, but this time I supported lift-over of vcf file which is interval data.


### Changes
- Interval end information (t-end q-end) has entered chain-data.
- Vcf-lift file is added.


### Tests
`lein test :all` :ok:
`lein check` :ok:
eastwood :ok:
cljfmt :ok:

alumi commented on src/varity/vcf_lift.clj:
```diff
@@ -0,0 +1,130 @@
+
+(defn- new-interval [variant chains]
+  (let [beg (:pos variant)
+        end (or (get-in variant [:info :end])
```
BTW, for nucleotide changes, we don't need to refer to the INFO/END but we have to update it in results.

alumi commented on src/varity/vcf_lift.clj:
```diff
@@ -0,0 +1,130 @@
+                (dec (+ beg (count (:ref variant)))))]
+    (->> chains
+         (ch/overlap-chains beg end)
+         (keep #(let [overlap-blocks (ch/overlap-blocks beg end (:data %))]
+                  (when-not (empty? overlap-blocks)
```
```suggestion
         (keep #(when-let [overlap-blocks (seq (ch/overlap-blocks beg end (:data %)))]
```
https://github.com/bbatsov/clojure-style-guide#nil-punning
https://github.com/bbatsov/clojure-style-guide#when-let

niyarin commented:
(I fixed it, but I forgot the comment.I'm sorry.)

・I used the chains index, so I reverted the chain.clj of my first commit and added necessary things.
・After that, I corrected the pointed out point.

alumi commented on src/varity/vcf_lift.clj:
```diff
@@ -0,0 +1,128 @@
+                                end-offset)
+                      strand (get-in % [:header :q-strand])]
+                  (if (= strand :reverse)
+                    {:start (- (get-in % [:header :q-size]) to-end)
+                     :end (- (get-in % [:header :q-size]) to-start)
+                     :strand strand}
+                    {:start to-start :end to-end :strand strand}))))
```
Also, `:start` and `:end` values seems to be off by one.
The lifted over region of `{:chr "chr1", :start 6, :end 6}` with `sample-rev.chain` should be `{:chr "chr1", :start 55, :end 55}` (both 1-based closed)

alumi commented on src/varity/vcf_lift.clj:
```diff
@@ -0,0 +1,128 @@
+                            info)
+                    :ref directed-ref
+                    :alt directed-alt)
+             (norm/realign seq-reader))))))
```
`shift-interval` seems to work so similarly to `norm/realign`. Wouldn't it be enough just to realign once? 

```clojure
(defn liftover-variant
  [rdr chain {:keys [chr pos ref info] :as variant}]
  (when-let [{:keys [chr start end strand]} (calc-new-interval
                                             chr pos (dec (+ pos (count ref)))
                                             chain)]
    (let [ref-seq (io-seq/read-sequence rdr {:chr chr, :start start, :end end})]
      (when (= ref-seq (cond-> ref (= strand :reverse) util-seq/revcomp))
        (cond-> (assoc variant :chr chr :pos start)
          (:END info) (assoc-in [:info :END] (dec (+ start (count ref))))
          (= strand :reverse) (-> (update :ref util-seq/revcomp)
                                  (update :alt #(map util-seq/revcomp %)))
          true (#(norm/realign rdr %)))))))
```

athos commented on src/varity/chain.clj:
```diff
@@ -122,3 +133,18 @@
+  "Calculats a list of blocks that overlap the given interval."
+  [start end ^Sorted blocks-idx]
+  (->> (subseq blocks-idx <= end)
+       (take-while (fn [[block-start block]]
+                     (<= block-start end)))
```
Here, `take-while` doesn't seem to do anything meaningful because `(subseq blocks-idx <= end)` ensures that all the returned entries have a key that is less than or equal to `end`. Probably, you need to do some more using `start` to cut down the head of the subsequence, right?

niyarin commented:
Thank you for pointing. I fixed them.

athos commented on src/varity/chain.clj:
```diff
@@ -122,3 +133,18 @@
+  "Calculates a list of blocks that overlap the given interval."
+  [start end ^Sorted blocks-idx]
+  (->> (subseq blocks-idx <= end)
+       (take-while (fn [[_ block]]
+                     (<= start (:t-end block))))
```
Should be `(drop-while (fn [[_ block]] (<= (:t-end block) start)))`? (I'm not so confident if `=` should really be included, though 😅)

And if it's correct, that implies a few more tests are necessary to capture the difference between the current implementation and the one I suggested above. I think it would be nice to have a couple of dedicated test cases for each `varity.chain/search-*` fn.

niyarin replied:
Thank you.My implementation is wrong.
`(drop-while (fn [[_ block]] (< (:t-end blcok)  start))`

niyarin commented:
I fixed Fixed the search function of chain.

niyarin commented:
I fixed searching chain and added test cases.

alumi commented:
[This simple code](https://github.com/chrovis/varity/pull/35#discussion_r460655826) still passes the test which means that either there're not enough test cases or the current implementation is needlessly complex.

alumi commented on test/varity/vcf_lift_test.clj:
```diff
@@ -0,0 +1,111 @@
+              (vcf/reader "./test-resources/vcf-lift/sample-multi.vcf")
+              seq-reader
+              (io-seq/reader "./test-resources/vcf-lift/sample.fasta")]
+    (is (= (vcf-lift/liftover-variants
+            seq-reader
+            (ch/index (ch/load-chain
+                       "./test-resources/vcf-lift/sample-multi.chain"))
+            (vcf/read-variants vcf-forward-rdr)))
+        {:failure nil,
+         :success [{:alt ["A"], :ref "ATGCATGC", :pos 9, :filter [:PASS],
+                    :id "1", :info {:END 16}, :qual 100.0, :chr "chr1"}
+                   {:alt ["A"], :ref "ATGCATGC", :pos 9, :filter [:PASS],
+                    :id "3", :info {:END 16}, :qual 100.0, :chr "chr1"}
+                   {:alt ["A"], :ref "ATGCATGCATGCATGCATGC", :pos 9,
+                    :filter [:PASS], :id "2", :info {:END 28}, :qual 100.0,
+                    :chr "chr1"}
+                   {:alt ["T"], :ref "TG", :pos 54, :filter [:PASS], :id "4",
+                    :info {:END 55}, :qual 100.0, :chr "chr1"}]})))
```
```suggestion
    (is (= (vcf-lift/liftover-variants
            seq-reader
            (ch/index (ch/load-chain
                       "./test-resources/vcf-lift/sample-multi.chain"))
            (vcf/read-variants vcf-forward-rdr))
           {:failure nil,
            :success [{:alt ["A"], :ref "ATGCATGC", :pos 9, :filter [:PASS],
                       :id "1", :info {:END 16}, :qual 100.0, :chr "chr1"}
                      {:alt ["A"], :ref "ATGCATGC", :pos 9, :filter [:PASS],
                       :id "3", :info {:END 16}, :qual 100.0, :chr "chr1"}
                      {:alt ["A"], :ref "ATGCATGCATGCATGCATGC", :pos 9,
                       :filter [:PASS], :id "2", :info {:END 28}, :qual 100.0,
                       :chr "chr1"}
                      {:alt ["T"], :ref "TG", :pos 54, :filter [:PASS], :id "4",
                       :info {:END 55}, :qual 100.0, :chr "chr1"}]}))))
```

niyarin replied:
I decided that simple code(https://github.com/chrovis/varity/pull/35#issuecomment-680415666) didn't work before, but it seems that I made a mistake when I checked.

I replaced with the simple code.
The result of throwing the Picard test was no problem.

alumi commented on src/varity/vcf_lift.clj:
```diff
@@ -0,0 +1,62 @@
+                :strand strand}
+               {:chr (get-in % [:header :q-name])
+                :start to-start :end to-end :strand strand})))
+        (ch/search-containing-chains chr beg end indexed-chains)))
```
I guess we need to compare scores if the query hits multiple chains.
Would you add tests for such cases?

niyarin commented:
I added a test case for hitting multiple chains.

alumi commented on src/varity/vcf_lift.clj:
```diff
@@ -0,0 +1,69 @@
+                              (update :alt #(map util-seq/revcomp %))))))
+
+(defn liftover-variant
+  "Lifts over a `variant` to another coordinate. Returns a converted and
+    realigned variant if succeeded to lift over, otherwise nil.
+      `indexed-chain` supplies conversions between sequences and must be indexed by
+        `varity.chain/index`. `target-seq-reader` is a reader instance of the target
+          sequence that the variant coordinate is converted to."
```
```suggestion
  "Lifts over a `variant` to another coordinate. Returns a converted and
  realigned variant if succeeded to lift over, otherwise nil.
  `indexed-chain` supplies conversions between sequences and must be indexed by
  `varity.chain/index`. `target-seq-reader` is a reader instance of the target
  sequence that the variant coordinate is converted to."
```

alumi commented on src/varity/vcf_lift.clj:
```diff
@@ -0,0 +1,69 @@
+      (norm/realign target-seq-reader x))))
+
+(defn liftover-variants
+  "Lifts over variants data from chains applied chain/index and seq-reader.
+   Variants that succeed in liftover will be put in :success,
+  and and those that failed will be put in :failure."
+  [seq-reader indexed-chains variants]
```
```suggestion
  "Lifts over variants data from chains applied chain/index and seq-reader.
  Variants that succeed in liftover will be put in :success,
  and and those that failed will be put in :failure."
  [seq-reader indexed-chains variants]
```

alumi commented on src/varity/chain.clj:
```diff
@@ -122,3 +125,18 @@
        (keep (partial in-block? pos))))
+
+(defn search-containing-chains
+  "Calculats a list of chains that contains the given interval."
```
```suggestion
  "Calculates a list of chains that contain the given interval."
```

alumi commented on src/varity/chain.clj:
```diff
@@ -122,3 +125,18 @@
+       (sort-by (comp :score :header) >)))
+
+(defn search-overlap-blocks
+  "Calculates a list of blocks that contains the given interval."
```
```suggestion
  "Calculates a list of blocks that overlap the given interval."
```

alumi reviewed (changes requested):
Thank you for the update! There're a few minor points that need to be fixed, but the others look good to me 👍

alumi reviewed (approved):
👍 👍 👍 
Please don't forget to squash your commits before marking this as ready for review.

alumi commented on src/varity/chain.clj:
```diff
@@ -122,3 +125,18 @@
+
+(defn search-overlap-blocks
+  "Calculates a list of blocks that overlap the given interval."
+  [start end ^Sorted blocks-idx]
```
```suggestion
  [start end blocks-idx]
```

athos commented on test/varity/vcf_lift_test.clj:
```diff
@@ -0,0 +1,124 @@
+      (is (= (#'vcf-lift/calc-new-interval "chr1" 191 192 chain-index)
+             {:chr "chr1", :start 51, :end 52, :strand :forward})))
+    (testing "ascending score chain"
+      (is (= (#'vcf-lift/calc-new-interval "chr4" 131 191 (ch/index (ch/load-chain "./test-resources/vcf-lift/sample-multi.chain")))
```
`chain-index` can be reused for this test case as well, right?

niyarin replied:
Thank you for pointing. I fixed it.

athos reviewed (commented):
Sorry for the terribly delayed response!

Just left a somewhat nitpicky comment. Other than that point, LGTM 👍

athos reviewed (approved):
Thanks for updating! Now let's move on to the varity team's review 💪

niyarin commented:
@totakke 
The Draft version of PR has been accepted by alumi and athos, so please review it.

I added liftover function for vcf files.

totakke commented:
@niyarin OK, thank you very much. I'm a little busy today, so I'll review it tonight or tomorrow.

totakke reviewed (commented):
Great job! Though I think there is no functional problem, I added two short comments.

niyarin commented:
Thank you for pointing,I fixed them.

totakke reviewed (approved):
LGTM

niyarin commented:
Thank you for reviewing.
